examples/hif2a/fit_to_multiple_rbfes.py

```diff
@@ -0,0 +1,431 @@
+        ("complex", complex_system, complex_coords, complex_box, conf.num_complex_windows),
+        ("solvent", solvent_system, solvent_coords, solvent_box, conf.num_solvent_windows)]:
+
+        lambda_schedule = construct_lambda_schedule(num_host_windows)
+
+        print("Minimizing the host structure to remove clashes...")
+        minimized_host_coords = minimizer.minimize_host_4d(rfe.mol_a, host_system, host_coords, rfe.ff, host_box)
+
+        # one GPU job per lambda window
+        print('submitting tasks to client!')
+        do_work = functools.partial(wrap_method, fxn=rfe.host_edge)
+        futures = []
+        for lambda_idx, lamb in enumerate(lambda_schedule):
+            arg = (lamb, host_system, minimized_host_coords, host_box, conf.num_equil_steps, conf.num_prod_steps)
+            futures.append(client.submit(do_work, arg))
+
+        results = []
+        for fut in futures:
+            results.append(fut.result())
+
+        stage_results[stage] = ThermodynamicIntegrationResult(lambda_schedule, results)
+
+        print(f'{stage} results by lambda window:')
+        for lamb, result in zip(lambda_schedule, results):
+            _print_result(lamb, result)
+
+        # estimate dG for this stage
+        pred_dG = np.trapz([_mean_du_dlambda(x) for x in results], lambda_schedule)
+        # TODO: refactor this to be a call to something in a generic free_energy.analysis module
+        #   to allow comparison / swapping between TI estimator, MBAR estimator, and others
+        stage_dGs[stage] = pred_dG
+
+        # collect derivative handlers
+        ghs = []
+        for lamb, result in zip(lambda_schedule, results):
+            bonded_du_dl, nonbonded_du_dl, grads_and_handles = result
+            ghs.append(grads_and_handles)
+
+        # update forcefield gradient handlers in-place
+        _update_combined_handle_and_grads(ghs, combined_handle_and_grads)
+
+    pred_rbfe = stage_dGs['complex'] - stage_dGs['solvent']
+    return pred_rbfe, combined_handle_and_grads, stage_results
+
+
+# TODO: define more flexible update rules here, rather than update parameters
+step_sizes = {
+    nonbonded.AM1CCCHandler: 1e-3,
+    nonbonded.LennardJonesHandler: 1e-3,
+    # ...
+}
+
+gradient_clip_thresholds = {
+    nonbonded.AM1CCCHandler: 0.001,
+    nonbonded.LennardJonesHandler: np.array([0.001, 0]),  # TODO: allow to update epsilon also?
+    # ...
+}
+
+
+def _clipped_update(gradient, step_size, clip_threshold):
+    """Compute an update based on current gradient
+        x[k+1] = x[k] + update
+
+    The gradient descent update would be
+        update = - step_size * grad(x[k]),
+
+    and to avoid instability, we clip the absolute values of all components of the update
+        update = - clip(step_size * grad(x[k]))
+
+    TODO: menu of other, fancier update functions
+    """
+    return - np.clip(step_size * gradient, -clip_threshold, clip_threshold)
```

Comment:
In practice, the gradient clipping will basically "binarize" the updates, the relative scales of the gradients are mostly removed.

Reply:
Multiplying the gradient estimate by a small `step_size` before applying the clipping operation means that the updates will not be binarized as often,  so the relative scales are retained more often.

As discussed off-line, on iterations where the gradient estimate can be flagged as highly unreliable (e.g. if it accompanies a deltaG estimate so large it might as well be NaN), we may also want to set the `step_size` or `clip_threshold` to 0. Setting the `clip_threshold` to a small number safeguards against having occasional nonsense gradients cause immediate training instability.